Unify `wgpu-core` and `wgpu-rs` types

GabrielMajeri opened the pull request:
**Connections**
Closes #689.

**Description**
Moves a lot of types from `wgpu-rs` which were duplicated in `wgpu-core` to `wgpu-types`.

**Testing**
Checked with core, player and `wgpu-rs`.

Corresponding `wgpu-rs` PR: https://github.com/gfx-rs/wgpu-rs/pull/437

kvark commented on player/src/main.rs:
```diff
@@ -273,7 +272,7 @@ impl GlobalExt for wgc::hub::Global<IdentityPassThroughFactory> {
                         label: Some(&label),
                         layout: layout_id,
-                        entries: &entry_vec,
+                        bindings: &entry_vec,
```
something bad happened here

kvark commented on wgpu-core/src/binding_model.rs:
```diff
@@ -260,29 +255,12 @@ pub struct BufferBinding {
-    pub binding: u32,
-    pub resource: BindingResource<'a>,
-}
+pub type Binding<'a> = wgt::Binding<BindingResource<'a>>;
```
we should call it `BindGroupEntry`, actually

kvark commented on wgpu-core/src/device/mod.rs:
```diff
@@ -1467,7 +1467,7 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         // Check that the number of entries in the descriptor matches
         // the number of entries in the layout.
-        let actual = desc.entries.len();
+        let actual = desc.bindings.len();
```
this looks like a backward change

kvark commented on wgpu-core/src/instance.rs:
```diff
@@ -25,23 +25,7 @@ use hal::{
-        }
-    }
-}
+pub type RequestAdapterOptions = wgt::RequestAdapterOptions<SurfaceId>;
```
nice to see so much stuff reused from wgpu-types! 🎉

kvark commented on wgpu-types/src/lib.rs:
```diff
@@ -1443,6 +1465,116 @@ impl<L> SamplerDescriptor<L> {
 
+/// Resource that can be bound to a pipeline.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum BindingResource<'a, Buffer, Sampler, TextureView, TextureViewArrayElement> {
```
ok, I think this is going a bit too far :)
Let's just have separate types here, we don't want to use more than a few generics

kvark commented on wgpu-types/src/lib.rs:
```diff
@@ -1443,6 +1465,116 @@ impl<L> SamplerDescriptor<L> {
+
+/// Bindable resource and the slot to bind it to.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Binding<R> {
```
we should call this `BindGroupEntry`

kvark commented on wgpu-types/src/lib.rs:
```diff
@@ -1443,6 +1465,116 @@ impl<L> SamplerDescriptor<L> {
+    /// The resources to bind to this bind group.
+    pub bindings: &'a [B],
+    /// Debug label of the bind group. This will show up in graphics debuggers for easy identification.
+    pub label: Option<&'a str>,
```
nit: can we move the label up to be the first field?

GabrielMajeri commented:
I've addressed the review comments.

I've also reviewed all of the `wgpu-rs` public types, and I think this should cover all of them (at least, all of those we can share between the crates without too many generics).
